BismarkUnchained/scenes/UI/Upgrades.gd

```diff
@@ -0,0 +1,89 @@
+extends Control
+
+var line
+var to_save = false
+
+# Passive upgrades
+var upgrades = 0
+var maxhealth = 100
+var regen = 0
+var lifesteal = 0
+var rof = 0.75
+
+# Upgrade values
+const rof_upgrade = 0.9
+const maxhealth_upgrade = 25
+const regen_upgrade = 2
+const lifesteal_upgrade = 2
+
+func _ready():
+	var save_file = File.new()
+	if save_file.file_exists("res://bismarkunchained.sav"):
+		save_file.open("res://bismarkunchained.sav", File.READ)
+		line = parse_json(save_file.get_line())
+		upgrades = line.upgrades
+		maxhealth = line.maxhealth
+		regen = line.regen
+		lifesteal = line.lifesteal
+		rof = line.rof
+		save_file.close()
+	
+	get_node("Labels/Upgrades").text = "Upgrades Available: " + str(upgrades)
+	get_node("Labels/RoFInfo").text = "Current RoF: " + str(rof)
+	get_node("Labels/MaxHealthInfo").text = "Current Max Health: " + str(maxhealth)
+	get_node("Labels/HealthRegenInfo").text = "Current Health Regen/sec: " + str(regen)
+	get_node("Labels/LifestealInfo").text = "Current Lifesteal: " + str(lifesteal) + "%"
+		
+
+func _on_RoF_pressed():
+	if upgrades > 0:
+		to_save = true
```

Comment:
see "save" comment: why not just save every time a change is made?

Reply:
we could do that, but I want a way for player to cancel changes if they want